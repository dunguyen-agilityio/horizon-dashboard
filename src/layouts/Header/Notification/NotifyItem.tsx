// Libs
import { memo } from 'react';

// Models
import { Notify } from '@/models/Notify';

// Components
import Text from '@/components/Text';

// Types
import { TEXT_SIZE } from '@/types/text';

// Utils
import { formatDateRelativeToNow } from '@/utils/format';

const NotifyItem = ({
  createdAt,
  message,
  unread,
}: Pick<Notify, 'message' | 'unread' | 'createdAt'>) => (
  <div className="flex gap-2 items-center">
    <div className="flex-1 flex flex-col">
      <Text size={TEXT_SIZE.md} className="font-bold">
        Reminder
      </Text>
      <Text size={TEXT_SIZE.sm} className="whitespace-break-spaces">
        {message}
      </Text>
      <Text size={TEXT_SIZE.xs}>{formatDateRelativeToNow(createdAt)}</Text>
    </div>
    <div className="w-10 pr-2">
      {unread && (
        <div className="bg-green-500 w-2 h-2 rounded-full float-end" />
      )}
    </div>
  </div>
);

export default memo(NotifyItem);
