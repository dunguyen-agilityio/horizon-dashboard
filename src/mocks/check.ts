import { TCheck } from '@/types/check';

export const MOCK_CHECKS: TCheck[] = [
  {
    id: '1',
    name: 'Tony Reichert',
    progress: 45,
    quantity: 20,
    createdAt: new Date('2024-08-14T08:25:04.188Z'),
  },
  {
    id: '2',
    name: 'Zoey Lang',
    progress: 15,
    quantity: 20,
    createdAt: new Date('2024-08-14T08:25:04.188Z'),
  },
  {
    id: '3',
    name: 'Jane Fisher',
    progress: 55,
    quantity: 28,
    createdAt: new Date('2024-08-14T08:25:04.188Z'),
  },
  {
    id: '4',
    name: 'William Howard',
    progress: 65,
    quantity: 30,
    createdAt: new Date('2024-08-14T08:25:04.188Z'),
  },
];
