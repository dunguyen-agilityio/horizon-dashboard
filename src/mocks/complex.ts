import { Complex, ComplexStatus } from '@/types/complex';

export const MOCK_COMPLEXES: Complex[] = [
  {
    id: '1',
    name: 'Tony Reichert',
    progress: 45,
    status: ComplexStatus.APPROVED,
    createdAt: new Date('2024-08-14T08:25:04.188Z'),
  },
  {
    id: '2',
    name: 'Zoey Lang',
    progress: 15,
    status: ComplexStatus.DISABLE,
    createdAt: new Date('2024-08-14T08:25:04.188Z'),
  },
  {
    id: '3',
    name: 'Jane Fisher',
    progress: 55,
    status: ComplexStatus.ERROR,
    createdAt: new Date('2024-08-14T08:25:04.188Z'),
  },
  {
    id: '4',
    name: 'William Howard',
    progress: 65,
    status: ComplexStatus.APPROVED,
    createdAt: new Date('2024-08-14T08:25:04.188Z'),
  },
];
