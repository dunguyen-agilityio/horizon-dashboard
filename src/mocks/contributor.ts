import { Contributor } from '@/models/Contributor';
import { MOCK_USERS } from './user';

export const MOCK_CONTRIBUTORS = [
  {
    ...MOCK_USERS[0],
    rating: 23,
    template: 'Abstract Colors',
    createdAt: new Date('2024-01-13T07:25:04.188Z'),
  },
  {
    ...MOCK_USERS[1],
    rating: 80,
    template: 'Swipe Circles',
    createdAt: new Date('2024-08-13T07:25:04.188Z'),
  },
  {
    ...MOCK_USERS[2],
    template: 'Colorful Heaven',
    createdAt: new Date('2024-08-11T07:25:04.188Z'),
  },
  {
    ...MOCK_USERS[3],
    template: '3D Cubes Art',
    avatar: undefined,
  },
] as Contributor[];
