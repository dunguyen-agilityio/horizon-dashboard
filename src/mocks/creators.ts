export const MOCK_CREATORS = [
  {
    id: '1',
    avatar: '/default-avatar.webp',
    userName: 'maddison_c21',
    artwork: 9821,
    rating: 90,
  },
  {
    id: '2',
    avatar: '/default-avatar.webp',
    userName: 'karl.willO2',
    artwork: 7032,
    rating: 80,
  },
  {
    id: '3',
    avatar: '/default-avatar.webp',
    userName: 'andreea.1z',
    artwork: 5204,
    rating: 60,
  },
];
