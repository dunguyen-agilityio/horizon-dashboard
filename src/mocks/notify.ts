import { Notify } from '@/models/Notify';

// Type
import { INotification } from '@/types/notification';

export const MOCK_NOTIFIES: Notify[] = [
  new Notify({
    id: '1',
    message: 'Your task is scheduled to start in 10 minutes.',
    createdAt: new Date('2024-08-14T08:25:04.188Z'),
    link: '/kanban/1',
    unread: true,
  }),
  new Notify({
    id: '2',
    message: 'Your task is scheduled to start in 10 minutes.',
    createdAt: new Date('2024-08-13T07:25:04.188Z'),
    link: '/kanban/2',
    unread: true,
  }),
  new Notify({
    id: '3',
    message: 'Your task is scheduled to start in 10 minutes.',
    createdAt: new Date('2024-08-14T02:25:07.188Z'),
    link: '/kanban/3',
  }),
  new Notify({
    id: '4',
    message: 'Your task is scheduled to start in 10 minutes.',
    createdAt: new Date('2024-08-08T08:25:04.188Z'),
    link: '/kanban/4',
  }),
  new Notify({
    id: '5',
    message: 'Your task is scheduled to start in 10 minutes.',
    createdAt: new Date('2024-08-08T08:25:04.188Z'),
    link: '/kanban/5',
  }),
  new Notify({
    id: '6',
    message: 'Your task is scheduled to start in 10 minutes.',
    createdAt: new Date('2024-08-08T08:25:04.188Z'),
    link: '/kanban/6',
  }),
  new Notify({
    id: '7',
    message: 'Your task is scheduled to start in 10 minutes.',
    createdAt: new Date('2024-08-08T08:25:04.188Z'),
    link: '/kanban/7',
  }),
  new Notify({
    id: '8',
    message: 'Your task is scheduled to start in 10 minutes.',
    createdAt: new Date('2024-08-08T08:25:04.188Z'),
    link: '/kanban/8',
  }),
];

export const CONTROL_NOTIFICATIONS: INotification[] = [
  {
    id: '1',
    isSelected: true,
    isDisabled: false,
    label: 'Item update notifications',
  },
  {
    id: '2',
    isSelected: false,
    isDisabled: true,
    label: 'Item comment notifications',
  },
  {
    id: '3',
    isSelected: false,
    isDisabled: true,
    label: 'Buyer review notifications',
  },
  {
    id: '4',
    isSelected: false,
    isDisabled: true,
    label: 'Rating reminders notifications',
  },
  {
    id: '5',
    isSelected: false,
    isDisabled: true,
    label: 'Meetups near you notifications',
  },
  {
    id: '6',
    isSelected: false,
    isDisabled: true,
    label: 'Company news notifications',
  },
  {
    id: '7',
    isSelected: false,
    isDisabled: true,
    label: 'New launches and projects',
  },
  {
    id: '8',
    isSelected: false,
    isDisabled: true,
    label: 'Monthly product changes',
  },
  {
    id: '9',
    isSelected: false,
    isDisabled: true,
    label: 'Subscribe to newsletter',
  },
  {
    id: '10',
    isSelected: true,
    isDisabled: true,
    label: 'Email me when someone follows me',
  },
];
