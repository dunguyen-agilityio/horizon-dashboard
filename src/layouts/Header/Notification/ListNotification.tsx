'use client';

// Libs
import { cn } from '@nextui-org/theme';

// Components
import { Listbox, ListboxItem } from '@nextui-org/listbox';

// Models
import { Notify } from '@/models/Notify';

// Styles
import '@/styles/scroll.css';
import NotifyItem from './NotifyItem';

interface ListNotificationProps {
  notifies: Notify[];
  onSelect: (key: React.Key) => void;
}

const ListNotification = ({ notifies, onSelect }: ListNotificationProps) => (
  <div data-testid="notification-list">
    <Listbox
      classNames={{
        base: 'max-w-md',
        list: 'max-h-[500px] overflow-y-scroll scrollbar pr-2',
      }}
      items={notifies}
      label="Assigned to"
      variant="flat"
      onAction={onSelect}
    >
      {({ link, id, unread, ...rest }) => (
        <ListboxItem
          key={link}
          textValue={id}
          className={cn(
            'border-b-secondary dark:border-b-indigo border-b-1 data-[hover=true]:bg-default/20',
            unread ? 'bg-default/20' : 'rounded-b-none hover:rounded-b-small',
          )}
          data-testid="notification-item"
        >
          <NotifyItem unread={unread} {...rest} />
        </ListboxItem>
      )}
    </Listbox>
  </div>
);

export default ListNotification;
