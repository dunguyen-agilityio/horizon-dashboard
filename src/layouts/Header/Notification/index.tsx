import { useRouter } from 'next/navigation';

// Components
import NotifyIcon from '@/icons/Notify';
import ListNotification from './ListNotification';
import { Popover, PopoverContent, PopoverTrigger } from '@nextui-org/react';

// Models
import { Notify } from '@/models/Notify';

interface NotificationProps {
  notifies: Notify[];
}

const Notification = ({ notifies }: NotificationProps) => {
  const { push } = useRouter();

  const handleSelect = (key: React.Key) => {
    push(key as string);
  };

  return (
    <Popover
      placement="bottom"
      showArrow={true}
      className="before:bg-gray data-[placement=bottom]:before:left-[64%] before:dark:bg-indigo-light"
    >
      <PopoverTrigger>
        <button
          className="focus-visible:outline-none aria-expanded:scale-1 aria-expanded:opacity-100"
          data-testid="notify-button"
        >
          <NotifyIcon />
        </button>
      </PopoverTrigger>
      <PopoverContent className="bg-gray dark:bg-indigo-light min-w-[320px] items-start w-full pr-[2px]">
        <ListNotification notifies={notifies} onSelect={handleSelect} />
      </PopoverContent>
    </Popover>
  );
};

export default Notification;
