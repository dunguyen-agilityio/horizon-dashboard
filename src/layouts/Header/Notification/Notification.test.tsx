import { fireEvent, render, waitFor } from '@testing-library/react';
import Notification from '.';
import { MOCK_NOTIFIES } from '@/mocks/notify';

const mockPush = jest.fn();

jest.mock('next/navigation', () => ({
  useRouter: jest.fn().mockImplementation(() => ({
    push: mockPush,
  })),
}));

describe('Notification tests', () => {
  const setup = () => render(<Notification notifies={MOCK_NOTIFIES} />);
  it('Should match snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('Should show Notification list when click button', async () => {
    const { getByTestId, queryByTestId } = setup();
    expect(queryByTestId('notification-list')).not.toBeInTheDocument();

    fireEvent.click(getByTestId('notify-button'));
    await waitFor(() => {
      expect(queryByTestId('notification-list')).toBeInTheDocument();
    });
  });

  it('Should trigger action link when user click Notification Item', () => {
    const { getByTestId, getAllByTestId } = setup();

    fireEvent.click(getByTestId('notify-button'));

    const { link } = MOCK_NOTIFIES[0];
    fireEvent.click(getAllByTestId('notification-item')[0]);
    expect(mockPush).toHaveBeenNthCalledWith(1, link);
  });
});
