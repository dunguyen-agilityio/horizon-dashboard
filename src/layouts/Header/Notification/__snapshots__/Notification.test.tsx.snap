// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification tests Should match snapshot 1`] = `
<div>
  <button
    aria-expanded="false"
    class="z-10 subpixel-antialiased focus-visible:outline-none aria-expanded:scale-1 aria-expanded:opacity-100"
    data-slot="trigger"
    data-testid="notify-button"
    type="button"
  >
    <i
      class="w-6 h-6 flex justify-center items-center hover:cursor-pointer fill-secondary dark:fill-white"
    >
      <svg
        height="20"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.29 15.29 14 14V9c0-3.07-1.64-5.64-4.5-6.32V2C9.5 1.17 8.83.5 8 .5S6.5 1.17 6.5 2v.68C3.63 3.36 2 5.92 2 9v5L.71 15.29c-.63.63-.19 1.71.7 1.71h13.17c.9 0 1.34-1.08.71-1.71ZM12 15H4V9c0-2.48 1.51-4.5 4-4.5s4 2.02 4 4.5v6Zm-4 5c1.1 0 2-.9 2-2H6a2 2 0 0 0 2 2Z"
        />
      </svg>
    </i>
  </button>
</div>
`;
